util/realistic_data/ingest_echo_data.py

```diff
@@ -64,13 +81,34 @@ def main():
     hdf_page_iterator = echo.paginate_hdf_data()
     total_ingestion_start_time = time()
 
+    last_successful_file = Config.config.script_options.file_to_restart_ingestion
+    ingestion_restarted = False if last_successful_file else True
+
     for page in hdf_page_iterator:
         object_names = [hdf_file["Key"] for hdf_file in page["Contents"]]
         page_start_time = time()
 
-        for name in object_names:
-            hdf_file_dict = echo.download_hdf_file(name)
-            og_api.submit_hdf(hdf_file_dict)
+        if last_successful_file in object_names:
+            print(f"Last successful file found: {last_successful_file}")
+            ingestion_restarted = True
+
+        if ingestion_restarted:
+            if Config.config.script_options.ingest_mode == "sequential":
+                for name in object_names:
```

Comment:
If `last_successful_file` is set then we skip the `page`, which makes sense. But we iterate over all the `object_names` - should we skip these as well until we find `last_successful_file`? Is there a guarantee that `last_successful_file` is the first entry in the page?

Maybe something like:
```python
object_names = object_names[object_names.index(last_successful_file):]
```
between lines 92 and 93?

Reply:
When ingestion is done in parallel, you need to re-ingest the entire page because you can't be sure which files were successfully ingested and which weren't. So yes, the current behaviour means that some files are ingested twice (which doesn't change the outcome of Echo/MongoDB, just takes slightly longer) but it means that no files are missed.

Reply:
OK, so `last_successful_file` might not actually be the last successful file, it's basically any file from the last page of object_names, since the user's probably just copying something from an error log and in principle there could be N files that we're being ingested in parallel when it fell over which may or may not have succeeded? Regardless of whether that's an fair description or not if we don't need to go through the whole page file by file that's fine.